nautobot_golden_config/utilities/config_to_push.py

```diff
@@ -0,0 +1,162 @@
+"""Functions related to prepare configuration to push."""
+from functools import partial
+from typing import Optional
+import jinja2
+from jinja2 import exceptions as jinja_errors
+
+from django.http import HttpRequest
+from django.core.exceptions import ObjectDoesNotExist
+
+from nautobot.dcim.models import Device
+from nautobot.users.models import User
+from nautobot.extras.models.secrets import SecretsGroup
+from nautobot.extras.choices import SecretsGroupAccessTypeChoices
+
+from netutils.utils import jinja2_convenience_function
+
+from nautobot_golden_config import models
+from nautobot_golden_config.utilities.constant import PLUGIN_CFG, ENABLE_PUSH
+from nautobot_golden_config.utilities.graphql import graph_ql_query
+from nautobot_golden_config.exceptions import RenderConfigToPushError
+from nautobot_golden_config.utilities.helper import get_device_to_settings_map
+
+
+def get_secret_by_secret_group_slug(
+    user: User,
+    secrets_group_slug: str,
+    secret_type: str,
+    secret_access_type: Optional[str] = SecretsGroupAccessTypeChoices.TYPE_GENERIC,
+) -> Optional[str]:
+    """Gets the secret from a Secret Group slug. To be used as a Jinja filter.
+
+    We assume that there is only one secret group corresponding to a model.
+
+    Args:
+        user (User): User object that performs API call to render push template with secrets.
+        secrets_group_slug (str): Secrets Group slug. It needs to be part of the GraphQL query.
+        secret_type (str): Type of secret, such as "username", "password", "token", "secret", or "key".
+        secret_access_type (Optional[str], optional): Type of secret such as "Generic", "gNMI", "HTTP(S)". Defaults to "Generic".
+
+    Returns:
+        Optional[str] : Secret value. None if there is no match. An error string if there is an error.
+    """
+    try:
+        secrets_group = SecretsGroup.objects.get(slug=secrets_group_slug)
+    except ObjectDoesNotExist:
+        return f"{secrets_group_slug} doesn't exist."
+
+    if not user.has_perm("extras.view_secretsgroup", secrets_group):
+        return f"You have no permission to read this secret {secrets_group_slug}."
+
+    return secrets_group.get_secret_value(
+        access_type=secret_access_type,
+        secret_type=secret_type,
+    )
+
+
+def _get_device_agg_data(device, request):
+    """Helper method to retrieve GraphQL data from a device."""
+    settings = get_device_to_settings_map(Device.objects.filter(pk=device.pk))[device.id]
+    _, device_data = graph_ql_query(request, device, settings.sot_agg_query.query)
```

Comment:
Trying to validate for the intended generation with config_to_push we call graphql twice?

Reply:
yes, I'm calling it to retrieve the data, but it could take the latest retrieved output if it was stored in some place. is it?

Reply:
This could reduce "a lot" the overall time this feature takes

Reply:
is there action required here?

Reply:
@chadell ping

Reply:
mmm, an improvement idea could be store graphql result, when doing the intended render (I understand it is not available), and reuse it if available, to same some second on the rendering. But maybe it would overcomplicate for now?
No problem to do if you think it's necessary

Reply:
I will leave it to you, honestly just making sure everything is addressed.